README.md
# vue-pipeline

[![npm](https://img.shields.io/npm/v/vue-pipeline.svg)](https://www.npmjs.com/package/vue-pipeline)
[![npm](https://img.shields.io/npm/dt/vue-pipeline.svg)](https://www.npmjs.com/package/vue-pipeline)
[![GitHub stars](https://img.shields.io/github/stars/jinfang134/vue-pipeline.svg?style=social&label=Stars&style=for-the-badge)](https://github.com/jinfang134/vue-pipeline/stargazers)
[![GitHub forks](https://img.shields.io/github/forks/jinfang134/vue-pipeline.svg?style=social&label=Fork&style=for-the-badge)](https://github.com/jinfang134/vue-pipeline/network)
[![license](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/jinfang134/vue-pipeline/blob/master/LICENSE)
[![Build Status](https://travis-ci.com/jinfang134/vue-pipeline.svg?branch=master)](https://travis-ci.com/jinfang134/vue-pipeline)

One easy-to-use component to show beautiful responsive timeline like jenkins blue ocean plugin.

[中文](https://github.com/jinfang134/vue-pipeline/blob/master/docs/cn.md)

![sample](https://github.com/jinfang134/vue-pipeline/raw/master/resources/sample.png)

To get started, check out:

- [Demo](https://jinfang134.github.io/vue-pipeline/)

## LICENSE

**NOTE:** Vue Pipeline is licensed under [The MIT License](https://github.com/jinfang134/vue-pipeline/blob/master/LICENSE). Completely free, you can arbitrarily use and modify this plugin. If this plugin is useful to you, you can **Star** this repo, your support is my biggest motive force, thanks.

## Features

- Created Graph according your data dynamiclly
- Responsive web design
- svg component
- Fully configurable
- Via data attributes
- Show/Hide arrow
- 3 kinds of lines
- support graph and tree view
- Single node selection
- Different status for each node
- Different weight for each edge
- Different color for each node and edge

## Install

```
npm install vue-pipeline
```

```
import Vue from 'vue'
import VuePipeline from 'vue-pipeline'

Vue.use(VuePipeline)

```

## Props

### Props of Pipeline

| Name      | Type    | Default | Description                                            |
| --------- | ------- | ------- | ------------------------------------------------------ |
| x         | number  | 50      | The x coordinate of the starting point of the graph    |
| y         | number  | 55      | The y coordinate of the starting point of the graph    |
| xstep     | number  | 120     | The position horizontally from a previous node.        |
| ystep     | number  | 50      | The position vertically from a previous node.          |
| data      | Array   | []      | data                                                   |
| lineStyle | string  | default | There are 3 types of line: ' default',' bessel','line' |
| showArrow | boolean | false   | whether show arrow for each line.                      |
|           |         |         |                                                        |

### Props for each node

| Name         | Type   | Default | Description                                                                                                      |
| ------------ | ------ | ------- | ---------------------------------------------------------------------------------------------------------------- |
| name         | string | null    | The title of each node                                                                                           |
| hint         | string | null    | The hint of each node                                                                                            |
| status       | string | null    | Status of each node, There are 6 type of status: `start`,`succeed`,`running`,`failure`,`paused`,`unstable`,`end` |
| next         | Array  | []      | The edge connected with this node                                                                                |
| next: index  | number | null    | The index of another node of this edge                                                                           |
| next: weight | number | null    | The weight of this edge                                                                                          |

**Sample Data:**

```javascript
let data = [
  {
    name: "Start",
    hint: "1m23s",
    status: "start",
    next: [{ index: 1, weight: 2 }]
  },
  {
    name: "Ammouncement Import",
    hint: "1m23s",
    status: "success",
    next: [
      { index: 2, weight: 0 },
      { index: 4, weight: 2 }
    ]
  },
  {
    name: "Employee ID to Onboarding",
    hint: "2m23s",
    status: "failure",
    next: [{ index: 3, weight: 0 }]
  },
  {
    name: "Personal Basic Info",
    hint: "2m23s",
    status: "paused",
    next: [{ index: 4, weight: 0 }]
  },
  { name: "End ", hint: "2m23s", status: "end", next: [] }
];
```

## Events

| Name   | Params | Description                 |
| ------ | ------ | --------------------------- |
| @click | node   | Occurs when node is clicked |

## Function

| Name   | params | Description                  |
| ------ | ------ | ---------------------------- |
| render | Node   | render the whole graph again |

## Contributing

If you find any bugs and/or want to contribute, feel free to create issues or submit pull requests.

Thanks!

## Local Development

```
yarn install
// Compiles and hot-reloads for development
yarn run serve
```

### publish a new version

```
npm version patch
npm version minor
npm version major
```

### Lints and fixes files

```
yarn run lint
```


changelog.md
# Changelog Vue 3



Files: 

/index.js => New (maybe incomplete) way to use a plugin in Vue 3. I only tested it on localhost.

/.eslintrc.js => Switching to Vue 3 (and using standard eslint because it was more comfortable for me personally). 

```json
extends: [
    'plugin:vue/vue3-essential',
    '@vue/standard'
  ],
  parserOptions: {
    ecmaVersion: 2020
  },
```

/src/components

service.js and line.js=> minor changes. Removal of `var` and change to `const` or `let` as appropriate, removal of unused variables in jsdoc and change of == to === (type checking in addition to the value).

*.vue => structurally there are no changes. Only the format has been changed from Vue 2 to Vue 3 respecting everything that existed before.



/demo/

main.js and App.vue=> change to Vue 3.



In package.json several dependencies were changed but I still have to adjust some things. If I can I will review this weekend the tests.


demo/data.js
const hue3 = {
  nodes: [
    {
      name: 'Start',
      hint: '1m23s',
      status: 'start',
      next: [{ index: 1, weight: 2 }]
    },
    {
      name: 'Ammouncement Import',
      hint: '1m23s',
      status: 'success',
      next: [
        { index: 2, weight: 0 },
        { index: 3, weight: 2 }
      ]
    },
    {
      name: 'Employee ID to Onboarding',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 3, weight: 0 }]
    },
    {
      name: 'Personal Basic Info',
      hint: '2m23s',
      status: 'paused',
      next: [{ index: 4, weight: 0 }]
    },
    { name: 'End ', hint: '2m23s', status: 'end', next: [] }
  ]
}

const hue1 = {
  nodes: [
    {
      name: 'Start',
      hint: '1m23s',
      status: 'start',
      next: [{ index: 1, weight: 2 }]
    },
    {
      name: 'Ammouncement Import',
      hint: '1m23s',
      status: 'success',
      next: [{ index: 2, weight: 2 }]
    },
    {
      name: 'Employee ID to Onboarding',
      hint: '2m23s',
      status: 'success',
      next: [
        { index: 3, weight: 2 },
        { index: 4, weight: 2 },
        { index: 5, weight: 2 },
        { index: 6, weight: 0 },
        { index: 7, weight: 1 },
        { index: 8, weight: 2 },
        { index: 9, weight: 0 }
      ]
    },
    {
      name: 'Personal Basic Info',
      hint: '2m23s',
      status: 'unstable',
      next: [{ index: 10, weight: 2 }]
    },
    {
      name: '地址信息',
      hint: '2m23s',
      status: 'success',
      next: [{ index: 10, weight: 0 }]
    },
    {
      name: 'Family Info',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 10, weight: 0 }]
    },
    {
      name: 'Education Info',
      hint: '2m23s',
      status: 'running',
      next: [{ index: 10, weight: 0 }]
    },
    {
      name: 'Degree Info',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 10, weight: 0 }]
    },
    {
      name: 'Career Info',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 10, weight: 0 }]
    },
    {
      name: 'Qualification Info',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 10, weight: 0 }]
    },
    { name: 'End', hint: '2m23s', status: 'end', next: [] }
  ]
}

const hue2 = {
  nodes: [
    {
      name: 'Start',
      hint: '1m23s',
      status: 'success',
      next: [
        {
          index: 1,
          weight: 0
        }
      ]
    },
    {
      name: 'Ammouncement Import',
      hint: '1m23s',
      status: 'success',
      next: [2, 3, 4, 5, 6]
    },
    {
      name: 'Personal Basic Info',
      hint: '2m23s',
      status: 'unstable',
      next: [4]
    },
    { name: 'Address Info', hint: '2m23s', status: 'success', next: [7] },
    { name: 'Family Info', hint: '2m23s', status: 'failure', next: [7] },
    { name: 'Family In', hint: '2m23s', status: 'failure', next: [7] },
    { name: 'Education Info', hint: '2m23s', status: 'success', next: [7] },
    { name: 'Degree Info', hint: '2m23s', status: 'paused', next: [8] },
    { name: 'End', hint: '2m23s', status: 'failure' }
  ]
}

const sample = {
  nodes: [
    {
      name: 'test0',
      hint: '1m23s',
      status: 'success',
      next: [{ index: 1, weight: 2 }]
    },
    {
      name: 'test1 long',
      hint: '1m23s',
      status: 'success',
      next: [
        { index: 2, weight: 2 },
        { index: 3, weight: 2 },
        { index: 4, weight: 2 }
      ]
    },
    {
      name: 'test2 hello world',
      hint: '2m23s',
      status: 'unstable',
      next: [{ index: 5 }]
    },
    {
      name: 'test3',
      hint: '2m23s',
      status: 'success',
      next: [{ index: 7, weight: 1 }]
    },
    {
      name: 'test4 ha',
      hint: '2m23s',
      status: 'running',
      next: [{ index: 8 }, { index: 9 }]
    }, // 4
    { name: 'test5', hint: '2m23s', status: 'failure', next: [{ index: 6 }] },
    { name: 'test6', hint: '2m23s', status: 'success', next: [{ index: 10 }] }, // 6
    {
      name: 'test7 hello',
      hint: '2m23s',
      status: 'paused',
      next: [{ index: 12 }, { index: 13 }]
    },
    {
      name: 'test8 2344',
      hint: '2m23s',
      status: 'paused',
      next: [{ index: 14 }, { index: 15 }]
    }, // 8
    { name: 'test9', hint: '2m23s', status: 'failure', next: [{ index: 16 }] },
    { name: 'test10', hint: '2m23s', status: 'failure', next: [{ index: 11 }] }, // 10
    { name: 'test11', hint: '2m23s', status: 'failure' },
    { name: 'test12', hint: '2m23s', status: 'failure' },
    { name: 'test13', hint: '2m23s', status: 'failure' },
    { name: 'test14', hint: '2m23s', status: 'failure' },
    { name: 'test15', hint: '2m23s', status: 'failure' },
    { name: 'test16', hint: '2m23s', status: 'failure' }
  ]
}

const sample3 = {
  nodes: [
    {
      name: 'test0',
      hint: '1m23s',
      status: 'success',
      next: [{ index: 1, weight: 2 }]
    },
    {
      name: 'test1 long',
      hint: '1m23s',
      status: 'success',
      next: [
        { index: 2, weight: 2 },
        { index: 3, weight: 2 },
        { index: 4, weight: 2 }
      ]
    },
    {
      name: 'test2 hello world',
      hint: '2m23s',
      status: 'unstable',
      next: [{ index: 5 }]
    },
    {
      name: 'test3',
      hint: '2m23s',
      status: 'success',
      next: [{ index: 7, weight: 1 }]
    },
    {
      name: 'test4 ha',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 8 }, { index: 9 }]
    }, // 4
    { name: 'test5', hint: '2m23s', status: 'failure', next: [{ index: 6 }] },
    { name: 'test6', hint: '2m23s', status: 'success', next: [{ index: 10 }] }, // 6
    {
      name: 'test7 hello',
      hint: '2m23s',
      status: 'paused',
      next: [{ index: 10 }]
    },
    {
      name: 'test8 2344',
      hint: '2m23s',
      status: 'paused',
      next: [{ index: 10 }]
    }, // 8
    { name: 'test9', hint: '2m23s', status: 'failure', next: [{ index: 10 }] },
    { name: 'test10', hint: '2m23s', status: 'failure', next: [{ index: 11 }] }, // 10
    { name: 'test11', hint: '2m23s', status: 'failure' }
  ]
}
const sample2 = {
  nodes: [
    {
      name: 'test0',
      hint: '1m23s',
      status: 'success',
      next: [{ index: 1, weight: 2 }],
      x: 50,
      y: 55
    },
    {
      name: 'test1 long',
      hint: '1m23s',
      status: 'success',
      next: [
        { index: 2, weight: 2 },
        { index: 3, weight: 2 },
        { index: 4, weight: 2 }
      ],
      x: 170,
      y: 55
    },
    {
      name: 'test2 hello world',
      hint: '2m23s',
      status: 'unstable',
      next: [{ index: 5 }],
      x: 290,
      y: 55
    },
    {
      name: 'test3',
      hint: '2m23s',
      status: 'success',
      next: [{ index: 7, weight: 1 }],
      x: 290,
      y: 125
    },
    {
      name: 'test4 ha',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 8 }, { index: 9 }],
      x: 290,
      y: 55
    },
    {
      name: 'test5',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 6 }],
      x: 410,
      y: 55
    },
    {
      name: 'test6',
      hint: '2m23s',
      status: 'running',
      next: [{ index: 10 }],
      x: 530,
      y: 55
    },
    {
      name: 'test7 hello',
      hint: '2m23s',
      status: 'paused',
      next: [{ index: 12 }, { index: 13 }, { index: '10' }],
      x: 650,
      y: 55
    },
    {
      name: 'test8 2344',
      hint: '2m23s',
      status: 'paused',
      next: [{ index: 14 }, { index: 15 }],
      x: 410,
      y: 55
    },
    {
      name: 'test9',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 16 }],
      x: 410,
      y: 265
    },
    {
      name: 'test10',
      hint: '2m23s',
      status: 'failure',
      next: [{ index: 11 }, { index: '11' }],
      x: 770,
      y: 55
    },
    { name: 'test11', hint: '2m23s', status: 'failure', x: 890, y: 55 },
    { name: 'test12', hint: '2m23s', status: 'failure', x: 530, y: 125 },
    { name: 'test13', hint: '2m23s', status: 'failure', x: 530, y: 195 },
    {
      name: 'test14',
      hint: '2m23s',
      status: 'failure',
      x: 530,
      y: 55,
      next: [{ index: '7' }]
    },
    {
      name: 'test15',
      hint: '2m23s',
      status: 'failure',
      x: 530,
      y: 335,
      next: [{ index: '7' }]
    },
    { name: 'test16', hint: '2m23s', status: 'failure', x: 530, y: 405 }
  ]
}

module.exports = {
  hue3,
  hue2,
  hue1,
  sample,
  sample2,
  sample3,
  bug: [
    { name: '0', next: [{ index: 1 }, { index: 3 }] },
    { name: '1', next: [{ index: 2 }, { index: 3 }] },
    { name: '2', next: [] },
    { name: '3', next: [] }
  ]
}
// export default hue1
